[UNIT 2] re-release evaluation unit for 2025 (#231)
huggingface/smol-course: units/en/unit2/2.md

vLLM (pronounced "vee-LLM") is a high-throughput and memory-efficient inference engine for large language models. Originally developed in the Sky Computing Lab at UC Berkeley, vLLM has evolved into a community-driven project that revolutionizes LLM serving with several key innovations:

### Key Features

- State-of-the-art serving throughput with PagedAttention for efficient memory management
- Continuous batching of incoming requests for optimal GPU utilization
- Multiple parallelism strategies: tensor, pipeline, data, and expert parallelism
- Quantization support: GPTQ, AWQ, INT4, INT8, and FP8
- OpenAI-compatible API server for easy deployment

### Why vLLM Matters for Evaluation

When running large-scale evaluations, traditional inference methods become bottlenecks:
- Memory waste: Traditional attention mechanisms store KV cache inefficiently
- Sequential processing: Cannot batch requests of different lengths effectively
- GPU underutilization: Frequent memory transfers leave GPUs idle

vLLM addresses these issues with revolutionary techniques that can achieve **up to 24x higher throughput** compared to traditional methods.

## VLLM and Hugging Face Integration

> vLLM integrates deeply with the Hugging Face ecosystem, as detailed in the [VLLM Hugging Face integration documentation](https://docs.vllm.ai/en/v0.9.0/design/huggingface_integration.html). Here's how it works:

Reviewer: ```suggestion
vLLM integrates deeply with the Hugging Face ecosystem, as detailed in the [VLLM Hugging Face integration documentation](https://docs.vllm.ai/en/latest/design/huggingface_integration.html). Here's how it works:
```

## Installation and Setup

### Basic Installation

```bash
# Install VLLM
pip install vllm

# Or with specific CUDA version
pip install vllm --extra-index-url https://download.pytorch.org/whl/cu121
```

### Verify Installation

```python
import vllm
print(f"VLLM version: {vllm.__version__}")
```

## Basic VLLM Usage with Hugging Face Models

### Simple Text Generation

```python
from vllm import LLM, SamplingParams

# Initialize the model
llm = LLM(
    model="HuggingFaceTB/SmolLM3-3B",
    trust_remote_code=True,
    dtype="float16",
    max_model_len=4096,
)

# Create sampling parameters
sampling_params = SamplingParams(
    temperature=0.6,
    top_p=0.95,
    max_tokens=512,
)

# Generate text
prompts = [
    "Explain the concept of machine learning in simple terms:",
    "Write a Python function to calculate factorial:",
]

outputs = llm.generate(prompts, sampling_params)

# Print results
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt}")
    print(f"Generated: {generated_text}")
    print("-" * 50)
```

<details>
<summary>Output</summary>

> <p>
> ```

Reviewer: ```suggestion
<p>

```
```

Prompt: Explain the concept of machine learning in simple terms:
Generated: Explain the concept of machine learning in simple terms:
--------------------------------------------------
Prompt: Write a Python function to calculate factorial:
Generated: Write a Python function to calculate factorial:
--------------------------------------------------

> ```
> </p>

Reviewer: ```suggestion
```

</p>
```

Looks like this at the moment:
<img width="1030" height="133" alt="Screenshot 2025-09-15 at 17 15 41" src="https://github.com/user-attachments/assets/407065e4-4eaf-4983-8a4b-17ed19469db8" />

```

<details>
<summary>Output</summary>
<p>

```
Give me a brief explanation of gravity in simple terms.
Gravity is the force that pulls objects towards each other. It is caused by the mass of the objects. The more mass an object has, the stronger its gravity is.
```

</p>
</details>

## VLLM API Server

For production deployments, VLLM provides an OpenAI-compatible API server:

### Starting the Server

```bash
# Basic server
vllm serve HuggingFaceTB/SmolLM3-3B \
    --dtype float16 \
    --trust-remote-code

# Multi-GPU server
vllm serve HuggingFaceTB/SmolLM3-3B \
    --tensor-parallel-size 4 \
    --dtype float16 \
    --trust-remote-code \
    --port 8000
```

### Using the API

```python
import openai

# Configure client
client = openai.OpenAI(
    base_url="http://localhost:8000/v1",
    api_key="token-abc123",  # Can be any string for local server
)

# Chat completion
response = client.chat.completions.create(
    model="HuggingFaceTB/SmolLM3-3B",
    messages=[
        {"role": "user", "content": "Explain quantum computing briefly."}
    ],
    temperature=0.6,
    max_tokens=512,
)

print(response.choices[0].message.content)
```

<details>
<summary>Output</summary>

> <p>
> ```

Reviewer: ```suggestion
<p>

```
```

Quantum computing is a type of computing that uses quantum bits instead of classical bits. Quantum bits are different from classical bits because they can be in a superposition of 0 and 1, which allows them to be in multiple states at once. This allows quantum computers to perform certain calculations much faster than classical computers.

> ```
> </p>

Reviewer: ```suggestion
```

</p>
```
